_drafts/monaderror.md

```diff
@@ -0,0 +1,236 @@
+---
+layout: post
+title: "Handling monadic errors"
+categories: learn fp
+---
+
+In the [last post] we have seen how our very own typeclasses can come in handy, and how we can use them in combination with Monads.
+I will assume you have read that.
+
+In this post, we will explore the options of error handling during monadic computations.
+
+## Motivation
+
+Our API was
+
+```scala
+trait Database[F[_]] {
+  def load(id: Int): F[User]
+  def save(user: User): F[Unit]
+}
+
+object Database {
+  object syntax {
+    def save[F[_]](user: User)(implicit db: Database[F]): F[Unit] = db.save(user)
+    def load[F[_]](id: Int)(implicit db: Database[F]): F[User] = db.load(id)
+  }
+}
+```
+
+We can implement an instance for any type constructor, and the problem is we know nothing here about errors. The different type constructors have different ways to approach errors, or none at all. Think about how different it is to deal with a failed `Future` from dealing with a `Left` for example. What can we do about it?
+
+One way, is to use `Either`, for example. Note, that in this post we are not abstracting over the type which represents the failure for the sake of simplicity, we will just stick with `Throwable`.
+
+```scala
+trait Database[F[_]] {
+  def load(id: Int): F[Either[Throwable, User]]
+  def save(user: User): F[Either[Throwable, Unit]]
+}
+
+object Database {
+  object syntax {
+    def load[F[_]](id: Int)(implicit db: Database[F]): F[Either[Throwable, User]] = db.load(id)
+    def save[F[_]](user: User)(implicit db: Database[F]): F[Either[Throwable, Unit]] = db.save(user)
+  }
+}
+```
+
+It is obviously more complex, but at least we have a unified way to deal with errors. 
+
+```scala
+import cats.Monad
+import cats.syntax.functor._
+import Database.syntax._
+
+def updateUser[F[_] : Database : Monad](userId: Int, newName: String): F[Either[Throwable, User]] = {
+  load(userId) map {
+    case Right(user) =>
+      val updated = user.copy(name = newName)
+      save(updated)
```

Comment:
note: as @gszabo pointed out, here a failure could go unnoticed, will be fixed soon